File: src/backend/histories_storage.py
from abc import ABC
import numpy as np
from typing import List, Tuple
from .trackable import Trackable


class HistoriesStorage:
    def __init__(self):
        self.objects = {}
        self.histories = {}
        
    def register_object(self, obj: Trackable):
        assert issubclass(type(obj), Trackable)
        if (obj_name := obj.display_name) in self.objects:
            # Check if the existing object is of the same type
            existing_obj = self.objects[obj_name]
            if type(existing_obj) == type(obj):
                object_type = type(obj).__name__
                raise RuntimeError(f'Duplicate {object_type}: An object with the name "{obj_name}" has already been registered.')
            else:
                existing_type = type(existing_obj).__name__
                new_type = type(obj).__name__
                raise RuntimeError(f'Name conflict: "{obj_name}" is already used by a {existing_type}, cannot use it for a {new_type}. Please ensure all objects have unique names.')
        else:
            self.objects[obj_name] = obj
            for field_name in obj.TRACKABLE_FIELDS:
                if not hasattr(obj, field_name):
                    raise ValueError(f'An error while trying to register an object "{obj_name}" with Histories. '
                                     f'The object doesn\'t have the "{field_name}" attribute.')
                self.histories[f'{obj_name}_{field_name}'] = []
        
    def update_histories(self):
        for obj_name, obj in self.objects.items():
            current_state = obj.get_current_state()
            for field_name, field_value in current_state.items():
                self.histories[f'{obj_name}_{field_name}'].append(field_value)
                
    def flush_histories(self):
        for tracked_field_name in self.histories.keys():
            self.histories[tracked_field_name] = []

    def reset(self):
        """Reset the storage to its initial state, clearing all objects and histories."""
        self.objects = {}
        self.histories = {}

    def display_histories(self):
        """Display histories in a structured format (for debugging only)."""
        # This method is only for debugging and doesn't need to print to terminal
        return {key: len(values) for key, values in self.histories.items()}
        
    def get_histories(self):
        return self.histories    

File: cpp_backend/src/HistoriesStorage.cpp
#include "HistoriesStorage.h"
#include "Vesicle.h"
#include "Exterior.h"
#include "IonSpecies.h"
#include "IonChannel.h"
#include <stdexcept>

HistoriesStorage::HistoriesStorage() {
    // Initialize empty maps
}

void HistoriesStorage::registerObject(std::shared_ptr<Trackable> obj) {
    if (!obj) {
        throw std::invalid_argument("Cannot register null object");
    }
    
    std::string objName = obj->getDisplayName();
    
    // Check for duplicates
    if (objects_.find(objName) != objects_.end()) {
        std::string existingType = typeid(*objects_[objName]).name();
        std::string newType = typeid(*obj).name();
        
        if (existingType == newType) {
            throw std::runtime_error("Duplicate object: An object with the name \"" + 
                                    objName + "\" has already been registered.");
        } else {
            throw std::runtime_error("Name conflict: \"" + objName + 
                                    "\" is already used by a " + existingType + 
                                    ", cannot use it for a " + newType + 
                                    ". Please ensure all objects have unique names.");
        }
    }
    
    // Register the object
    objects_[objName] = obj;
    
    // Initialize histories for each field in the object's current state
    auto currentState = obj->getCurrentState();
    for (const auto& [fieldName, fieldValue] : currentState) {
        std::string historyKey = objName + "_" + fieldName;
        histories_[historyKey] = std::vector<double>();
    }
}

void HistoriesStorage::registerObject(std::shared_ptr<Vesicle> obj) {
    // Create a Trackable adapter for Vesicle
    std::shared_ptr<Trackable> trackable = std::static_pointer_cast<Trackable>(obj);
    registerObject(trackable);
}

void HistoriesStorage::registerObject(std::shared_ptr<Exterior> obj) {
    // Create a Trackable adapter for Exterior
    std::shared_ptr<Trackable> trackable = std::static_pointer_cast<Trackable>(obj);
    registerObject(trackable);
}

void HistoriesStorage::registerObject(std::shared_ptr<IonSpecies> obj) {
    // Create a Trackable adapter for IonSpecies
    std::shared_ptr<Trackable> trackable = std::static_pointer_cast<Trackable>(obj);
    registerObject(trackable);
}

void HistoriesStorage::registerObject(std::shared_ptr<IonChannel> obj) {
    // Create a Trackable adapter for IonChannel
    std::shared_ptr<Trackable> trackable = std::static_pointer_cast<Trackable>(obj);
    registerObject(trackable);
}

void HistoriesStorage::updateHistories() {
    // Update histories with current state of all objects
    for (const auto& [objName, obj] : objects_) {
        auto currentState = obj->getCurrentState();
        for (const auto& [fieldName, fieldValue] : currentState) {
            std::string historyKey = objName + "_" + fieldName;
            histories_[historyKey].push_back(fieldValue);
        }
    }
}

void HistoriesStorage::flushHistories() {
    // Clear all history values but keep registered objects
    for (auto& [historyKey, historyValues] : histories_) {
        historyValues.clear();
    }
}

void HistoriesStorage::reset() {
    // Clear everything
    objects_.clear();
    histories_.clear();
}

const std::unordered_map<std::string, std::vector<double>>& HistoriesStorage::getHistories() const {
    return histories_;
}

nlohmann::json HistoriesStorage::toJson() const {
    // Convert histories to JSON
    nlohmann::json result;
    
    for (const auto& [historyKey, historyValues] : histories_) {
        result[historyKey] = historyValues;
    }
    
    return result;
}

void HistoriesStorage::addHistory(const std::string& name, double value) {
    histories_[name].push_back(value);
} 